enhancements/installer/signal-cluster-deletion.md

```diff
@@ -0,0 +1,130 @@
+---
+title: signal-cluster-deletion
+authors:
+  - "@abutcher"
+reviewers:
+  - dgoodwin
+  - abhinavdahiya
+approvers:
+  - dgoodwin
+  - abhinavdahiya
+creation-date: 2020-03-30
+last-updated: 2020-04-06
+status: provisional
+---
+
+# Signal Cluster Deletion
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Operators can create external resources in clouds and other services that have no opportunity to be cleaned up when a cluster is uninstalled, causing extra work for cluster maintainers to track down and remove these resources. Signalling cluster deletion would allow operators to clean up external resources during uninstallation and would ensure that total cleanup is an automatic process.
+
+## Motivation
+
+### Goals
+
+Signal that cluster deletion has begun through deletion of a new CRD `ClusterAlive` and wait for that object to be deleted during uninstall before continuing with removing cluster resources such as instances, storage and networking.
+
+### Non-Goals
+
+* Attaching finalizers to operator resources based on the `ClusterAlive` object which would facilitate removal of operator resources.
+* Removing resources other than the `ClusterAlive` object during cluster uninstall.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints
+
+#### API
+
+Introduce a new CRD `ClusterAlive` from which operator resources could attach a finalizer initiating cascading delete when the `ClusterAlive` resource is deleted.
+
+```yaml
+apiVersion: v1
+kind: ClusterAlive
+metadata:
+  name: cluster
+spec:
+  blockTeardown: true  # block teardown of operator resources
+```
+
+#### Install
+
+Create a `ClusterAlive` object during installation. (Is a manifest the right place to create for new clusters ?)
+
+#### Upgrade
+
+Create `ClusterAlive` object for existing clusters. (Is this the right place to create for existing clusters ?)
+
+#### Uninstall
+
+Delete a cluster's `ClusterAlive` resource during uninstall when requested via flag such as `openshift-install destroy cluster --cluster-alive-delete`. Cluster destroy will wait for a default amount of time and fail if `ClusterAlive` deletion was not successful. The default timeout will not be configurable and users will be expected to attempt shutdown multiple times upon failure.
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+In OpenShift CI, clusters will try to delete their `ClusterAlive` object and if that fails, all infra resources will be destroyed.
```

Comment:
How do I get a record of every time the `--cluster-alive-delete` failed?

Reply:
What do you think about a field in `.spec` or `.status` that we increment when we attempt the destroy?

Reply:
I meant in CI, how do I get that record? how do we know it's broken in CI and not just getting ignored by the ||true forever

Reply:
> ...how do we know it's broken in CI...

We have lots of sub-components in the installer that are not independently tested, so I don't think we need to block on closing this one.  If you consider it especially vulnerable, are you worried about the installer breaking or a particular component forgetting to register a finalizer or flubbing the finalizer implementation?  If you're worried about the installer, you could have CI ask a trusted-enough operator to create an unknown-to-the-installer resource which would block teardown, and then see if the backstopping reapers light up with lots of leaked resources.

Reply:
the `|| true` would only hit the case of a flubbed implementation from something setting a finalizer, which would be a real bug.

Your suggestion, I think, is that the aws reaper might eventually notice we were wasting resources/money and we could then do some spelunking to figure out what left it behind.

Could we somehow, instead do (obviously psuedocode)

retval =  openshift-install destroy cluster --cluster-alive-delete
if retval != success
    openshift-install destroy cluster
return retval

This moves the failure to where it is introduced, not to some point later after DPP realizes we're wasting money.


I do agree, we don't have to block closing on this issue. But it doesn't seem that hard.

Reply:
Updated test plan to include returning the retval of the cluster alive delete test if it failed.